src/sio.c

```diff
@@ -0,0 +1,259 @@
+/* SPDX-License-Identifier: MIT */
+
+#include "adt.h"
+#include "errno.h"
+#include "malloc.h"
+#include "sio.h"
+#include "types.h"
+#include "utils.h"
+
+// Reuse pages for different data sections, if space allows it
+#define MERGE_SIO_FWDATA
+
+#define SIO_KEY(s)  _SIO_KEY(#s)
+#define _SIO_KEY(s) (((s)[0] << 24) | ((s)[1] << 16) | ((s)[2] << 8) | (s)[3])
+
+#define MAX_FWDATA   8
+#define MAX_FWPARAMS 16
+
+int sio_num_fwdata;
+struct sio_mapping *sio_fwdata;
+int sio_num_fwparams;
+struct sio_fwparam *sio_fwparams;
+
+static void *alloc_mapped_data(size_t size, u64 *iova)
+{
+    if (sio_num_fwdata >= MAX_FWDATA)
+        return NULL;
+
+    struct sio_mapping *mapping = &sio_fwdata[sio_num_fwdata];
+
+#ifdef MERGE_SIO_FWDATA
+    if (sio_num_fwdata && ALIGN_UP((mapping - 1)->size, SZ_16K) >= (mapping - 1)->size + size) {
+        mapping--;
+        *iova = mapping->iova + mapping->size;
+        mapping->size = ALIGN_UP(mapping->size + size, SZ_4K);
```

Comment:
not sure if strictly necessary but in the merge case the data is not cleared. easiest workaround would be to always clear the whole 16k page:
`memset64((void *)mapping->phys, 0, ALIGN_UP(mapping->size, SZ_16K));`

Reply:
Good catch!